src/crypto/CHIPCryptoPALmbedTLS.cpp

```diff
@@ -504,83 +504,301 @@ void ClearSecretData(uint8_t * buf, uint32_t len)
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    int result       = 0;
+
+    memset(&context, 0, sizeof(context));
+    mbedtls_ecp_group_init(&context.curve);
+    result = mbedtls_ecp_group_load(&context.curve, MBEDTLS_ECP_DP_SECP256R1);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+
+    context.md_info = mbedtls_md_info_from_type(MBEDTLS_MD_SHA256);
+    VerifyOrExit(context.md_info != NULL, error = CHIP_ERROR_INTERNAL);
+
+    mbedtls_ecp_point_init(&context.M);
+    mbedtls_ecp_point_init(&context.N);
+    mbedtls_ecp_point_init(&context.X);
+    mbedtls_ecp_point_init(&context.Y);
+    mbedtls_ecp_point_init(&context.L);
+    mbedtls_ecp_point_init(&context.V);
+    mbedtls_ecp_point_init(&context.Z);
+    M = &context.M;
+    N = &context.N;
+    X = &context.X;
+    Y = &context.Y;
+    L = &context.L;
+    V = &context.V;
+    Z = &context.Z;
+
+    mbedtls_mpi_init(&context.w0);
+    mbedtls_mpi_init(&context.w1);
+    mbedtls_mpi_init(&context.xy);
+    mbedtls_mpi_init(&context.tempbn);
+    w0     = &context.w0;
+    w1     = &context.w1;
+    xy     = &context.xy;
+    tempbn = &context.tempbn;
+
+    G     = &context.curve.G;
+    order = &context.curve.N;
+
+    return error;
+
+exit:
+    _log_mbedTLS_error(result);
+
+    FreeImpl();
+    return error;
 }
 
-void Spake2p_P256_SHA256_HKDF_HMAC::FreeImpl(void) {}
+void Spake2p_P256_SHA256_HKDF_HMAC::FreeImpl(void)
+{
+    mbedtls_ecp_point_free(&context.M);
+    mbedtls_ecp_point_free(&context.N);
+    mbedtls_ecp_point_free(&context.X);
+    mbedtls_ecp_point_free(&context.Y);
+    mbedtls_ecp_point_free(&context.L);
+    mbedtls_ecp_point_free(&context.Z);
+    mbedtls_ecp_point_free(&context.V);
+
+    mbedtls_mpi_free(&context.w0);
+    mbedtls_mpi_free(&context.w1);
+    mbedtls_mpi_free(&context.xy);
+    mbedtls_mpi_free(&context.tempbn);
+
+    mbedtls_ecp_group_free(&context.curve);
+}
 
 CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::Mac(const unsigned char * key, size_t key_len, const unsigned char * in, size_t in_len,
                                               unsigned char * out)
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    int result       = 0;
+
+    mbedtls_md_context_t hmac_ctx;
+    mbedtls_md_init(&hmac_ctx);
+    result = mbedtls_md_setup(&hmac_ctx, context.md_info, 1);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+
+    result = mbedtls_md_hmac_starts(&hmac_ctx, key, key_len);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+
+    result = mbedtls_md_hmac_update(&hmac_ctx, in, in_len);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+
+    result = mbedtls_md_hmac_finish(&hmac_ctx, out);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+
+    return error;
+
+exit:
+    _log_mbedTLS_error(result);
+
+    mbedtls_md_free(&hmac_ctx);
+    return error;
+}
+
+static inline int const_memcmp(const void * a, const void * b, size_t n)
```

Comment:
why not memcmp? please add a comment if standard memcmp is really unsuitable.

Reply:
memcmp is not constant time - while it's unlikely to be exploitable it's general best practice to use constant time functions for cryptographic functions (including hmac comparisons).

Reply:
ok, for some reason I thought 'const_' prefix was referencing arguments being const (which makes no sense as memcmp also has them const).

This info would be great to be in a comment.

Reply:
I'll rename it & leave a comment to make it more clear !

Reply:
@pan-apple ?